### PR TITLE
DI container implementation 😱

### PR DESCRIPTION
**Pull request by joshski**
Hey @aslakhellesoy here's my current thinking about how we can avoid some of the issues we see with roles versus user interfaces as this assembly pattern scales.

To summarise my half-baked thoughts:

* Every character (such as an instance of `Maker` or `Breaker`) has many _views_ of the system. Views enable the character to trigger commands _and_ provide a snapshot of the current state of the system. So, many views means a) many abilities which change over time and b) many snapshots of the system's current state, to use in assertions.

* Triggering a command may spawn a new view (such as when a child window is launched in a desktop app, or a character opens a browser for the first time). Or a command may mutate an existing view (such as when a user logs in: they can no longer see the log in page without logging out first). It may _also_ mutate other characters views. For example, when a new person enters the lobby, everyone's lobby view is updated, thereby enabling new interactions for any character with a lobby view.

* The character instance is therefore very dynamic. A character does not have a static set of "abilities" - their abilities change over time, depending on the views they can see, which is determined by the set of commands that have been performed thus far.

* An assembly is therefore a strategy for choosing between competing options. For example: a) when a character wants to "paste", should they press command+v, right-click, or select Edit>Paste? b) When a controller needs to persist a game, should it use a memory game store, or a database game store? 

Because there is so much dynamic behaviour, I think a simple solution to this problem is to use dynamic dependency injection via JavaScript `Proxy` instances. This happens to play quite nicely with destructuring constructors e.g. `constructor({ x, y }) {}` resolves components with the `x` and `y` roles.

**Update:**

Two other things I have been experimenting with in this branch:

* Abstract hypermedia-style controllers
* Push

### CODE REVIEW

File: interfaces/controllers/HomeController.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module.exports = class HomeController {
+  openApplication() {
+    return {
+      description: "home",
+      forms: [{ action: "startGameWithWord" }]
```
Comment (PaulCampbell) [inline interfaces/controllers/HomeController.js]: is "form" an appropriate word.  It feels a little tied to a html implementation

maybe `actions: [{ intent: 'startGameWithWord' }]`

In the same way params are knowledge encoded in the test or the html, this "method" thing is possibly just knowledge encoded in the html

File: features/step_definitions/steps.js [hunk 2]
```diff
@@ -1,39 +1,125 @@
 const { Given, When, Then } = require("cucumber");
 const assert = require("assert");
 
-Given("{Maker} has started a game with the word {string}", async function(
-  maker,
-  word
+Given(
+  "{player} has started a game with the word {string}",
+  async function makerHasStartedGameWithWord(maker, word) {
```
Comment (PaulCampbell) [inline features/step_definitions/steps.js:6]: can this be something like:

```
const game = await maker.openApplication()
await game.actions['startGame']({word})
```

Comment (PaulCampbell) [reply features/step_definitions/steps.js:6]: hmm. or maybe something like:

```
const game = await maker.openApplication()
const ui = new WhateverUi(game)
await ui.actions['startGame']({word})
```

Comment (joshski) [reply features/step_definitions/steps.js:6]: I don't think you want the steps to be aware of the `WhateverUi`. The same steps should work with different implementations of the same UI protocol

File: interfaces/controllers/HomeController.js (hunk 1 above)
Comment (joshski) [reply interfaces/controllers/HomeController.js]: Yeah I was definitely thinking of HTML i.e. `<form action="">` but that's probably a bad idea at this level.

**Pull request merged**